File: src/python/signalbird/messaging.py
"""Gönderim (Messaging) istemcisi — sunucu tarafı.

Takım API anahtarıyla (``sb_…``) e-posta/SMS/push gönderir, kişi ve liste
yönetir, kampanya açar, mesaj durumlarını okur.

Metot adları diğer dillerle birebir aynıdır; Python'da snake_case yazılır
(``send_email`` ↔ ``sendEmail``) — bir dilin kendi yazım geleneğini bozmak,
paritenin sağladığı kolaylıktan fazlasını götürür.

Sözleşme: docs/CONTRACT.md § 8
"""

from __future__ import annotations

from typing import Any, Iterator, Mapping, Optional, Sequence

from ._http import DEFAULT_BASE_URL, Result, SignalbirdError, Transport, seg

BULK_CHUNK = 1000


class SignalbirdMessaging:
    def __init__(
        self,
        api_key: str,
        base_url: Optional[str] = None,
        timeout: float = 15.0,
        throw_on_error: bool = False,
        debug: bool = False,
    ):
        if not api_key:
            raise SignalbirdError("Signalbird: api_key zorunlu.", 0, "NO_KEY")

        # Telsiz (`sbr_`) ya da uygulama (`sbw_pub_`) anahtarı buraya verilirse
        # her istek 401 döner; kurulumda söylemek haftalar sonra bulunacak
        # hatayı önler.
        if not api_key.startswith("sb_"):
            raise SignalbirdError(
                "Signalbird: gönderim istemcisi takım API anahtarı ister (sb_…).",
                0,
                "WRONG_KEY_TYPE",
            )

        self._http = Transport(api_key, base_url or DEFAULT_BASE_URL, timeout, throw_on_error, debug)

    # ── Gönderim ─────────────────────────────────────────────────────────

    def send_email(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/email/send", input)

    def send_sms(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/sms/send", input)

    def preview_sms(self, body: str) -> Result:
        """SMS parça/karakter hesabı — kota harcamaz."""
        return self._http.request("POST", "/v1/sms/preview", {"body": body})

    def send_push(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/push/send", input)

    # ── Kişiler ──────────────────────────────────────────────────────────

    def list_contacts(self, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", "/v1/contacts", None, query)

    def create_contact(self, contact: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/contacts", contact)

    def update_contact(self, id: Any, contact: Mapping[str, Any]) -> Result:
        return self._http.request("PATCH", f"/v1/contacts/{seg(id)}", contact)

    def delete_contact(self, id: Any) -> Result:
        return self._http.request("DELETE", f"/v1/contacts/{seg(id)}")

    def bulk_contacts(self, input: Mapping[str, Any]) -> Result:
        """1000'lik parçalar hâlinde SIRAYLA yükler.

        Paralel değil: aynı e-posta iki parçadaysa yarış olmasın. Bir parça
        başarısız olursa o noktada durulur ve o ana kadar biriken sayımlar
        döner — yarısı yüklenmiş bir listeyi "başarısız" diye atmak, hangi
        kişinin girdiğini bilmemek demektir.
        """
        contacts: Sequence[Mapping[str, Any]] = list(input.get("contacts") or [])
        rest = {key: value for key, value in input.items() if key != "contacts"}

        totals = {"imported": 0, "updated": 0, "skipped": []}

        if not contacts:
            return Result(ok=True, status=200, data=totals, code=None, message=None)

        for start in range(0, len(contacts), BULK_CHUNK):
            chunk = contacts[start : start + BULK_CHUNK]
            result = self._http.request("POST", "/v1/contacts/bulk", {**rest, "contacts": list(chunk)})

            if not result.ok:
                return Result(
                    ok=False,
                    status=result.status,
                    data=totals,
                    code=result.code,
                    message=result.message,
                )

            data = result.data if isinstance(result.data, dict) else {}
            totals["imported"] += int(data.get("imported") or 0)
            totals["updated"] += int(data.get("updated") or 0)
            totals["skipped"].extend(data.get("skipped") or [])

        return Result(ok=True, status=200, data=totals, code=None, message=None)

    # ── Listeler ─────────────────────────────────────────────────────────

    def list_contact_lists(self) -> Result:
        return self._http.request("GET", "/v1/contact-lists")

    def create_contact_list(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/contact-lists", input)

    def delete_contact_list(self, id: Any) -> Result:
        return self._http.request("DELETE", f"/v1/contact-lists/{seg(id)}")

    # ── Kampanyalar ──────────────────────────────────────────────────────

    def list_campaigns(self, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", "/v1/campaigns", None, query)

    def create_campaign(self, input: Mapping[str, Any]) -> Result:
        return self._http.request("POST", "/v1/campaigns", input)

    def get_campaign(self, id: Any) -> Result:
        return self._http.request("GET", f"/v1/campaigns/{seg(id)}")

    def cancel_campaign(self, id: Any) -> Result:
        return self._http.request("POST", f"/v1/campaigns/{seg(id)}/cancel")

    def list_campaign_messages(self, id: Any, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", f"/v1/campaigns/{seg(id)}/messages", None, query)

    def iterate_campaign_messages(
        self, id: Any, query: Optional[Mapping[str, Any]] = None
    ) -> Iterator[Mapping[str, Any]]:
        """Sayfa sayfa gezer; büyük kampanyada tüm listeyi belleğe almaz."""
        page = int((query or {}).get("page") or 1)

        while True:
            result = self.list_campaign_messages(id, {**(query or {}), "page": page})

            if not result.ok:
                return

            data = result.data if isinstance(result.data, dict) else {}
            rows = data.get("data") or []

            for row in rows:
                yield row

            last = data.get("last_page")

            if not rows or (last is not None and page >= int(last)):
                return

            page += 1

    # ── Mesajlar ─────────────────────────────────────────────────────────

    def list_messages(self, query: Optional[Mapping[str, Any]] = None) -> Result:
        return self._http.request("GET", "/v1/messages", None, query)

    def get_message(self, id: str) -> Result:
        return self._http.request("GET", f"/v1/messages/{seg(id)}")

File: src/go/signalbird/messaging.go
package signalbird

import (
	"context"
	"strings"
	"time"
)

// BulkChunk — toplu kişi yüklemede tek istekteki üst sınır.
const BulkChunk = 1000

// KeyConfig — takım anahtarıyla çalışan istemcilerin (Gönderim, Yönetim) ayarı.
type KeyConfig struct {
	// APIKey, takım API anahtarıdır (sb_…). Telsiz anahtarı burada çalışmaz.
	APIKey  string
	BaseURL string
	// Timeout boşsa 15 saniye — toplu kişi yükleme uzun sürebilir.
	Timeout      time.Duration
	ThrowOnError bool
	Debug        bool
}

// Messaging — Gönderim istemcisi: e-posta/SMS/push, kişi, liste, kampanya, mesaj.
type Messaging struct {
	http *transport
}

// NewMessaging kurar. sb_ dışı anahtar kurulum anında reddedilir: sessizce 401
// yiyip haftalar sonra fark edilmesin.
func NewMessaging(config KeyConfig) (*Messaging, error) {
	http, err := newKeyTransport(config)
	if err != nil {
		return nil, err
	}

	return &Messaging{http: http}, nil
}

func newKeyTransport(config KeyConfig) (*transport, error) {
	if config.APIKey == "" {
		return nil, ErrNoKey
	}

	if !strings.HasPrefix(config.APIKey, "sb_") {
		return nil, ErrWrongKeyType
	}

	timeout := config.Timeout
	if timeout == 0 {
		timeout = 15 * time.Second
	}

	return newTransport(config.APIKey, config.BaseURL, timeout, config.ThrowOnError, config.Debug), nil
}

// ── Gönderim ───────────────────────────────────────────────────────────

// SendEmail — class alanı (transactional | commercial) ZORUNLUDUR ve
// varsayılanı yoktur: hukuki kapı çağıranın elindedir.
func (m *Messaging) SendEmail(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/email/send", input, nil)
}

func (m *Messaging) SendSms(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/sms/send", input, nil)
}

// PreviewSms — parça/karakter hesabı; kota harcamaz.
func (m *Messaging) PreviewSms(ctx context.Context, body string) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/sms/preview", map[string]any{"body": body}, nil)
}

func (m *Messaging) SendPush(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/push/send", input, nil)
}

// ── Kişiler ────────────────────────────────────────────────────────────

func (m *Messaging) ListContacts(ctx context.Context, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/contacts", nil, query)
}

func (m *Messaging) CreateContact(ctx context.Context, contact map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/contacts", contact, nil)
}

func (m *Messaging) UpdateContact(ctx context.Context, id any, contact map[string]any) (Result, error) {
	return m.http.request(ctx, "PATCH", "/v1/contacts/"+seg(id), contact, nil)
}

func (m *Messaging) DeleteContact(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/contacts/"+seg(id), nil, nil)
}

// BulkContacts, girdiyi 1000'lik parçalara böler ve SIRAYLA gönderir.
//
// Paralel değil: aynı e-posta iki parçadaysa yarış olmasın. Bir parça
// başarısız olursa o noktada durulur ve o ana kadarki sonuç döner.
func (m *Messaging) BulkContacts(ctx context.Context, input map[string]any) (Result, error) {
	contacts := normalizeContacts(input["contacts"])

	if len(contacts) == 0 {
		return Result{OK: true, Status: 200}, nil
	}

	rest := map[string]any{}
	for key, value := range input {
		if key != "contacts" {
			rest[key] = value
		}
	}

	var last Result

	for start := 0; start < len(contacts); start += BulkChunk {
		end := start + BulkChunk
		if end > len(contacts) {
			end = len(contacts)
		}

		payload := map[string]any{"contacts": contacts[start:end]}
		for key, value := range rest {
			payload[key] = value
		}

		result, err := m.http.request(ctx, "POST", "/v1/contacts/bulk", payload, nil)
		if err != nil || !result.OK {
			return result, err
		}

		last = result
	}

	return last, nil
}

// ── Listeler ───────────────────────────────────────────────────────────

func (m *Messaging) ListContactLists(ctx context.Context) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/contact-lists", nil, nil)
}

func (m *Messaging) CreateContactList(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/contact-lists", input, nil)
}

func (m *Messaging) DeleteContactList(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "DELETE", "/v1/contact-lists/"+seg(id), nil, nil)
}

// ── Kampanyalar ────────────────────────────────────────────────────────

func (m *Messaging) ListCampaigns(ctx context.Context, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/campaigns", nil, query)
}

// CreateCampaign — buradan çıkan her ileti ZORUNLU commercial'dır.
func (m *Messaging) CreateCampaign(ctx context.Context, input map[string]any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/campaigns", input, nil)
}

func (m *Messaging) GetCampaign(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/campaigns/"+seg(id), nil, nil)
}

func (m *Messaging) CancelCampaign(ctx context.Context, id any) (Result, error) {
	return m.http.request(ctx, "POST", "/v1/campaigns/"+seg(id)+"/cancel", nil, nil)
}

func (m *Messaging) ListCampaignMessages(ctx context.Context, id any, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/campaigns/"+seg(id)+"/messages", nil, query)
}

// IterateCampaignMessages, sayfaları gezer ve her satır için fn çağırır.
// fn false dönerse gezinme durur. Büyük kampanyada tüm listeyi belleğe almaz.
func (m *Messaging) IterateCampaignMessages(ctx context.Context, id any, query Query, fn func(row map[string]any) bool) error {
	page := 1

	if query == nil {
		query = Query{}
	}

	for {
		query["page"] = page

		result, err := m.ListCampaignMessages(ctx, id, query)
		if err != nil || !result.OK {
			return err
		}

		var payload struct {
			Data     []map[string]any `json:"data"`
			LastPage int              `json:"last_page"`
		}

		if err := result.Into(&payload); err != nil {
			return err
		}

		for _, row := range payload.Data {
			if !fn(row) {
				return nil
			}
		}

		if len(payload.Data) == 0 || (payload.LastPage > 0 && page >= payload.LastPage) {
			return nil
		}

		page++
	}
}

// ── Mesajlar ───────────────────────────────────────────────────────────

func (m *Messaging) ListMessages(ctx context.Context, query Query) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/messages", nil, query)
}

func (m *Messaging) GetMessage(ctx context.Context, id string) (Result, error) {
	return m.http.request(ctx, "GET", "/v1/messages/"+seg(id), nil, nil)
}

// normalizeContacts, hem []map[string]any hem []any kabul eder.
//
// Çağıran çoğu zaman JSON'dan gelen []any taşır; yalnız birini kabul etmek
// listeyi sessizce BOŞ sayıp "0 kişi yüklendi" demek olurdu — hataların en
// kötüsü hiçbir şey söylemeyendir.
func normalizeContacts(value any) []any {
	switch typed := value.(type) {
	case []any:
		return typed
	case []map[string]any:
		rows := make([]any, 0, len(typed))
		for _, row := range typed {
			rows = append(rows, row)
		}

		return rows
	default:
		return nil
	}
}
